src/webhook/webhook.service.ts

```diff
@@ -0,0 +1,99 @@
+import { HttpService } from '@nestjs/axios';
+import {
+  ForbiddenException,
+  Injectable,
+  NotFoundException,
+  UnauthorizedException,
+} from '@nestjs/common';
+import { CreateWebhookRequest, FetchedWebhook, WebhookEvent } from './dto';
+import { WebhookRepository } from './webhook.repository';
+import { Recipe } from '@prisma/client';
+import Cryptr from 'cryptr';
+
+@Injectable()
+export class WebhookService {
+  constructor(
+    private readonly httpService: HttpService,
+    private readonly webhookRepository: WebhookRepository,
+  ) {}
+  cryptr = new Cryptr(process.env.CRYPTR_SECRET_KEY);
+
+  async createWebhook(
+    userId: number,
+    webhookData: CreateWebhookRequest,
+  ): Promise<void> {
+    const userWebhooks = await this.webhookRepository.getAllWebhooksByUserId(
+      userId,
+    );
+    if (userWebhooks.length >= +process.env.WEBHOOK_LIMIT) {
+      throw new ForbiddenException();
+    }
+
+    if (webhookData.token) {
+      const encryptedString = this.cryptr.encrypt(webhookData.token);
+      webhookData.token = encryptedString;
+    }
+
+    this.webhookRepository.createWebhook(userId, webhookData);
+  }
+
+  async deleteWebhook(userId: number, webhookId: number) {
+    const webhook = await this.webhookRepository.getWebhookById(webhookId);
+    if (!webhook) {
+      throw new NotFoundException();
+    }
+    if (userId !== webhook.userId) {
+      throw new UnauthorizedException();
+    }
+    this.webhookRepository.deleteUserWebhookById(webhookId);
+  }
+
+  async getWebhooksByUserId(userId: number): Promise<FetchedWebhook[]> {
+    const webhooks = await this.webhookRepository.getAllWebhooksByUserId(
+      userId,
+    );
+    return webhooks.map(({ userId, token, ...rest }) => {
+      return rest;
+    });
+  }
+
+  sendToWebhook(url: string, data: Recipe, token?: string, attempt = 0): void {
+    let decryptedToken: string;
+    if (token) {
+      decryptedToken = this.cryptr.decrypt(token);
+    }
+    const headersRequest = {
+      'Content-Type': 'application/json',
+      Authorization: `Bearer ${token}`,
+    };
+    this.httpService
+      .post(url, data, {
+        headers: headersRequest,
+      })
+      .subscribe({
+        error: () => {
+          const maxWaitingTime = 86400 * 1000;
+          const nextTryInSec = Math.min(Math.pow(2, attempt + 1) * 1000);
```

Comment:
hmm, so your retry policy is 2s, 4s, 8s, 16s, ..., ~68 minutes, ..., ~18h 12m. I kind of don't like it 😛 Let's use our learning time and implement the following retry policy:

Retry | Time
1 | 2 minutes after the failure
2 | 6 minutes after the previous retry
3 | 30 minutes after the previous retry
4 | 1 hour after the previous retry
5 | 5 hours after the previous retry
6 | 1 day after the previous retry
7 | 2 days after the previous retry

(you can add it to the acceptance criteria)